src/App.tsx

```diff
@@ -1,19 +1,167 @@
+  } = useContext(ProcessedContext);
+
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filtredBy, setFiltredBy] = useState(FilterOptions.ALL);
+
+  const getUserTodos = async () => {
+    if (user) {
+      try {
+        const todosFromServer = await getTodos(user.id);
+
+        setTodos(todosFromServer);
+      } catch {
+        setError(ErrorTypes.GET);
+      }
+    }
+  };
+
+  const addNewTodo = async (todo: TodoData) => {
+    try {
+      setError(ErrorTypes.NONE);
+      setIsAdding(true);
+      const tempTodo = { ...todo, id: 0 };
+
+      setTodos(prevTodos => [...prevTodos, tempTodo]);
+
+      await createTodo(todo);
+    } catch {
+      setError(ErrorTypes.ADD);
+    } finally {
+      setIsAdding(false);
+      getUserTodos();
+    }
+  };
+
+  const deleteCurrentTodo = async (todoId: number) => {
+    try {
+      setError(ErrorTypes.NONE);
+      setProcessedTodoIds((currentIDs) => [...currentIDs, todoId]);
+
+      await deleteTodo(todoId);
+      setTodos(currentTodos => {
+        return currentTodos.filter(todo => todoId !== todo.id);
+      });
+    } catch {
+      setError(ErrorTypes.DELETE);
+    } finally {
+      setProcessedTodoIds([]);
+    }
+  };
+
+  const updateCurrentTodo = async (
+    todoId: number,
+    dataToUpdate: Partial<Todo>,
+  ) => {
+    try {
+      setError(ErrorTypes.NONE);
+      setProcessedTodoIds((currentIDs) => [...currentIDs, todoId]);
+      setTodos((currentTodos) => {
+        const updatedTodo = currentTodos.find(el => el.id === todoId) || null;
+
+        if (updatedTodo) {
+          Object.assign(updatedTodo, dataToUpdate);
+        }
+
+        return currentTodos;
+      });
+
+      await updateTodo(todoId, dataToUpdate);
+    } catch {
+      setError(ErrorTypes.UPDATE);
+    } finally {
+      await getUserTodos();
+      setProcessedTodoIds([]);
+    }
+  };
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
-    }
+    getUserTodos();
+  }, []);
+
+  const filterBySelect = useCallback((
+    todosFromServer: Todo[],
+    option: string,
+  ) => {
+    return todosFromServer.filter(todo => {
+      switch (option) {
+        case FilterOptions.ACTIVE:
+          return todo.completed === false;
+
+        case FilterOptions.COMPLETED:
+          return todo.completed === true;
+
+        case FilterOptions.ALL:
+        default:
+          return true;
+      }
+    });
   }, []);
 
+  const visibleTodos = useMemo(() => {
+    return filterBySelect(todos, filtredBy);
+  }, [todos, filtredBy]);
+
+  const activeTodos = useMemo(() => {
+    return filterBySelect(todos, FilterOptions.ACTIVE);
+  }, [todos]);
+
+  const isClearNeeded = useMemo(() => {
+    return activeTodos.length !== visibleTodos.length;
+  }, [activeTodos, visibleTodos]);
+
+  const clearCompletedTodos = async () => {
+    const compTodos = todos
+      .filter(todo => todo.completed);
```

Comment:
i wrote it below: here is better to make completedTodos variable for these two logics